### PR TITLE
[ENG-755] implement auto-submits

### PR DESCRIPTION
**Pull request by dsokal**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary.
- [ ] I've tagged the changelog entry with `[EAS BUILD API]` if it's a part of a breaking change to EAS Build API (only possible when updating `@expo/eas-build-job` package).

# Why

This is the last part for https://linear.app/expo/issue/ENG-755/submitting-automatically-on-build-success
The PR adds `--submit` and `--submit-with-profile` params to `eas build` that allow for submitting automatically on build complete.

# How

- All parts were already almost in place. I basically copy-pasted some code from `eas submit`.
- I added two CLI params to `eas build`:
   - `--submit` - creates a submission that's waiting for the build to complete. The CLI can be killed (CTRL+C) as soon as the submission has been created. The user doesn't need to wait for the build to complete. When this flag is supplied, the submit profile with the same name as the build profile is picked up.
   - `--submit-with-profile` - works the same as `--submit` but allows for using another submit profile.
- Currently, auto-submits are not supported with the `--local` flag. I'll work on this in another PR.

# Test Plan

Tested manually.

Screenshot 2021-09-09 at 18 10 22

### CODE REVIEW

File: packages/eas-cli/src/commands/build/index.ts [hunk 1]
```diff
@@ -93,6 +112,17 @@ export default class Build extends EasCommand {
       default: false,
       description: 'Clear cache before the build',
     }),
+    submit: flags.boolean({
```
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:115]: maybe `--auto-submit`

File: packages/eas-cli/src/commands/build/index.ts [hunk 2]
```diff
@@ -39,6 +53,8 @@ interface RawBuildFlags {
   wait: boolean;
   'clear-cache': boolean;
   json: boolean;
+  submit: boolean;
+  'submit-profile'?: string;
```
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:57]:
```suggestion
  'submit-with-profile'?: string;
```

File: packages/eas-cli/src/commands/build/index.ts [hunk 3]
```diff
@@ -184,15 +271,18 @@ export default class Build extends EasCommand {
...
       skipProjectConfiguration: flags['skip-project-configuration'],
-      profile: flags['profile'],
+      profile,
       nonInteractive,
       local: flags['local'],
       wait: flags['wait'],
       clearCache: flags['clear-cache'],
       json: flags['json'],
+      submit: flags['submit'],
+      submitProfile: flags['submit-profile'] ?? profile,
```
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:285]:
```suggestion
      submitProfile: flags['submit-with-profile'] ?? profile,
```

File: packages/eas-cli/src/commands/build/index.ts [hunk 4]
```diff
@@ -166,9 +248,14 @@ export default class Build extends EasCommand {
     if (flags.json && !nonInteractive) {
       error('--json is allowed only when building in non-interactive mode', { exit: 1 });
     }
-    const requestedPlatform = await selectRequestedPlatformAsync(flags.platform);
 
+    const requestedPlatform = await selectRequestedPlatformAsync(flags.platform);
     if (flags.local) {
+      if (flags.submit || flags['submit-profile'] !== undefined) {
```
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:254]:
```suggestion
      if (flags.submit || flags['submit-with-profile'] !== undefined) {
```

**Review by wkozyra95 — APPROVED** [review]

File: packages/eas-cli/src/commands/build/index.ts (hunk 1 above)
Comment (dsokal) [reply packages/eas-cli/src/commands/build/index.ts:115]: @brentvatne what do you think?

Comment (brentvatne) [reply packages/eas-cli/src/commands/build/index.ts:115]: i talked with @ide and @EvanBacon about this a bit, and i think @wkozyra95's suggestion is pretty reasonable. "auto submit is clearer at a glance to me. makes it clear that another action is taking place" - @ide.

as i said in the original doc, we could bikeshed on this for days. i don't think any of the options feel like they fit perfectly but this seems like the best for now, we can always revisit if we have some other insight.

so, the flags could be: `--auto-submit` and `--auto-submit-with-profile`. if we use the names, only one of the flags should be supplied at any given time: `eas build -p ios --auto-submit --auto-submit-with-profile beta` wouldn't make sense (we can ignore the `--auto-submit` and give the profile precedence here). 

```
eas build -p ios --profile abcd --auto-submit-with-profile beta 
eas build -p ios --profile abcd --auto-submit
```

my main concern is how long the `--auto-submit-with-profile` name is. but let's roll with it to unblock.

Comment (dsokal) [reply packages/eas-cli/src/commands/build/index.ts:115]: TIL https://en.wiktionary.org/wiki/bikeshedding Thanks @brentvatne :D

Comment (dsokal) [reply packages/eas-cli/src/commands/build/index.ts:115]: > if we use the names, only one of the flags should be supplied at any given time
This is exactly how I implemented it. See https://github.com/expo/eas-cli/pull/603/files#diff-cd2e00f4d286a4714f8736e14c635f4c3b98c402367cd71a83e4cd743034bd79R119

**Review by barthap — APPROVED** [review]
👍

**Review by wkozyra95 — APPROVED** [review]

**Comment by wkozyra95** [comment]
could you check how it behaves when build fails? Just to make sure we don't print errors that submission failed when the build was at fault

**Comment by rmujica** [comment]
Hi all, `--auto-submit` is not passing ENV variables defined in the build step into the submit step. 

I manage a collection of Expo project, each with its own Appstore and Playstore page, from one codebase. I used to use the `--config` flag selecting the respective app.json; but migrated to app.config.js and ENV variable to select one specific config json.

So, before the latest EAS updates, I started using `APP=someapp eas build --options` and `APP=someapp eas submit --platform ios|android --store options...`; then the eas.json scheme changed to allow saving the eas cli options into the json file; I saved the APP ENV variable into the build step, but the submit step doesn't allow passing an ENV variable into it, even though the submit step (I assume) still reads the app.config.js to check the Expo project that the build is related to.

I thought that the `--auto-submit `option would pass the ENV variable into submit step, but that didn't happen. I checked that the correct behavior was achieved when executing the CLI with the ENV variable: `APP=someapp eas build --platform=android --profile=profile --auto-submit`

**Comment by dsokal** [comment]
@rmujica thanks for reporting but would you mind creating an issue instead?

**Pull request merged**